crates/wit-parser/src/decoding.rs

```diff
@@ -81,45 +139,83 @@ impl ComponentInfo {
+                        if is_implicit {
+                            externs.push((
+                                import.name.0.to_string(),
+                                Extern::Import(import.name.0.to_string()),
+                            ));
+                        } else {
+                            cur_package.externs.push((
+                                import.name.0.to_string(),
+                                Extern::Import(import.name.0.to_string()),
+                            ));
+                        }
                     }
                 }
-                Payload::ComponentExportSection(s) if depth == 1 => {
+                Payload::ComponentExportSection(s) => {
                     for export in s {
                         let export = export?;
-                        externs.push((
-                            export.name.0.to_string(),
-                            Extern::Export(DecodingExport {
-                                name: export.name.0.to_string(),
-                                kind: export.kind,
-                                index: export.index,
-                            }),
-                        ));
+                        if is_implicit {
+                            externs.push((
+                                export.name.0.to_string(),
+                                Extern::Export(DecodingExport {
+                                    name: export.name.0.to_string(),
+                                    kind: export.kind,
+                                    index: export.index,
+                                }),
+                            ));
+                        } else {
+                            cur_package.externs.push((
+                                export.name.0.to_string(),
+                                Extern::Export(DecodingExport {
+                                    name: export.name.0.to_string(),
+                                    kind: export.kind,
+                                    index: export.index,
+                                }),
+                            ));
+                        }
                     }
                 }
                 #[cfg(feature = "serde")]
-                Payload::CustomSection(s) if s.name() == PackageMetadata::SECTION_NAME => {
-                    if _package_metadata.is_some() {
-                        bail!("multiple {:?} sections", PackageMetadata::SECTION_NAME);
+                Payload::CustomSection(s) => {
+                    if s.name() == PackageMetadata::SECTION_NAME {
+                        _package_metadata = Some(PackageMetadata::decode(s.data())?);
+                        cur_package.package_metadata = _package_metadata.clone();
+                    } else if s.name() == "component-name" {
+                        if let KnownCustom::ComponentName(reader) = s.as_known() {
+                            let _packages = register_names(reader, &mut cur_package)?;
```

Comment:
I'm a bit hesitant to rely on a custom section for something so significant to the structure of the final WIT, would it be possible to avoid this inference of infer via other means?

Reply:
So I actually had similar thoughts when implementing... But I think that the way names of components are handled today is exclusively via custom sections?  Will investigate some more, but if we want something similar to the wat outlined above where the name of the component is tied to the package name, I'm not sure of another way to get the information that informs a package name out of the binary format.

Reply:
Documentation and `@since`/`@unstable` are definitely all handled with custom sections, but to me those are semi-ok to strip at any time where the name of a package is significant enough that you'd never want it stripped. I'm mostly wondering if this can be inferred from the exported items or something like that rather than having to be listed in the name section. An alternative could be to export the components themselves from the outer component where the export name has the full package name (of sorts at least)

Reply:
I guess there's maybe an edge case here where you could have an empty scoped package?  I think we could infer things from exports so long as there are exports, and as far as I know the only case where there wouldn't be is if the package is empty...  Would you be comfortable erroring on empty scoped packages then?  Or yeah, we could also do the alternative and export the components themselves...

Reply:
@lukewagner do you have thoughts on this point specifically? The question being how is the package name recovered when multiple packages are encoded in a component and one package is possibly empty.

Reply:
Or yeah... I think I like your idea of exporting the component, seems like the easiest way to deduce the name of the component without relying on the name section...  How the PR handles it at the moment, even if we can tell it's an explicit/scoped package from the way the exports are structured, I think we still need the name section to figured out the actual name of the package that should be printed from a binary (unless there's another mechanism I'm missing).  Also curious to hear @lukewagner's thoughts about if we exported the components.